### PR TITLE
Audit after join, Audit Queueing, Swagger UI

### PR DESCRIPTION
**Pull request by Xyaren**
* Add Swagger UI on / route
* Add Audit on join (prioritized)
* Enable queued audit processing
* Move audit functions to separate service
* allow cli arguments to be passed via ENV variables
* allow log level and filepath to be provided via args
* fixed some linter warnings
* fix bug that caused leftover guild icons on the server on deleting the guild
* slightly improved performance of audit for users that are in the Database but not in the Ts DB (flipped if-conditions)
* remove reinit of ts connection as this is working with listeners and "state" of the connection. (Name, Channel etc)

### CODE REVIEW

File: bot/connection_pool/__init__.py [hunk 1]
```diff
@@ -34,14 +34,14 @@ class Unhealthy(Expired):
     """Connection was unhealthy"""
 
 
-T = TypeVar('T')
+_T = TypeVar('_T')
```
Comment (ogrady) [inline bot/connection_pool/__init__.py:37]: Why this change?

File: bot/gwapi/facade.py [hunk 2]
```diff
@@ -70,16 +75,19 @@ def guild_search(guild_name: str) -> Optional[str]:
...
+@cached(cache=TTLCache(maxsize=32, ttl=300))  # cache clients for 5 min - creation takes quite long
 def account_get(api_key: str) -> Account:
     api = _create_client(api_key=api_key)
     return _check_error(api.account.get())
 
 
+@cached(cache=TTLCache(maxsize=32, ttl=300))  # cache clients for 5 min - creation takes quite long
 def characters_get(api_key: str) -> List[Character]:
     api = _create_client(api_key=api_key)
     return _check_error(api.characters.get(page="0", page_size=200))
```
Comment (ogrady) [inline bot/gwapi/facade.py:87]: What is the reasoning behind these paging boundaries?

File: bot/main.py [hunk 3]
```diff
@@ -41,87 +41,94 @@ def main():  #
...
+                            response = ts_facade.wait_for_event(timeout=config.bot_sleep_idle)
+                            event_type: str = response.event
+                            event_data = response.parsed[0]
+                            if event_type == 'notifytextmessage':
+                                if "msg" in event_data:
+                                    # text message
+                                    bot_instance.messageEventHandler(event_data)  # handle event
+                            elif event_type == 'notifycliententerview':
+                                if event_data["client_type"] == '1':  # server query client
+                                    continue
```
Comment (ogrady) [inline bot/main.py:116]: Is there a reason for this style, instead of 
```python
if event_data["client_type"] != '1':
    bot_instance.loginEventHandler(event_data)  # handle event
```
?

File: bot/rest/dist/static/openapi-spec.yaml [hunk 4]
```diff
@@ -3,9 +3,23 @@ info:
   title: ts-gw2-verifyBot
   description: ts-gw2-verifyBot
   version: 1.0.0
-servers:
-  - url: 'http://localhost:10137'
 paths:
+  /health:
+    get:
+      summary: Healthckeck
+      operationId: helathCheck
```
Comment (ogrady) [inline bot/rest/dist/static/openapi-spec.yaml:10]: There seems to be a typo here.

File: bot/ts/TS3Facade.py [hunk 5]
```diff
@@ -205,11 +205,17 @@ def client_ids_from_uid(self, client_uid):
...
+        return self._ts3_connection.force_rename(target_nickname=target_nickname)
 
     def channel_edit(self, channel_id: str, new_channel_name: str):
         return self._ts3_connection.ts3exec(lambda tsc: tsc.exec_("channeledit", cid=channel_id, channel_name=new_channel_name), signal_exception_handler)
 
+    def remove_icon_if_exists(self, icon_id: int):
+        icon_server_path: str = f"/icon_{icon_id}"
+        return self._ts3_connection.ts3exec(lambda tsc: tsc.exec_("ftdeletefile", cid=0, cpw=None, name=icon_server_path), ignore_exception_handler)
+
+    pass
```
Comment (ogrady) [inline bot/ts/TS3Facade.py:217]: Leftover `pass`?

**Review by ogrady — COMMENTED** [review]
Looks generally good, see some inline comments.

File: bot/connection_pool/__init__.py (hunk 1 above)
Comment (Xyaren) [reply bot/connection_pool/__init__.py:37]: Plyint wanted it to be a private variable.

File: bot/main.py (hunk 3 above)
Comment (Xyaren) [reply bot/main.py:116]: I found it much more readable as it's basically an explicit skipping of this event

Comment (ogrady) [reply bot/main.py:116]: I strongly oppose any use of `continue` (at all), `break` (outside of `switch`), and `return` (unless as very last statement of a function or as early exit at the start of a function), which are all [`goto`s in disguise](https://homepages.cwi.nl/~storm/teaching/reader/Dijkstra68.pdf). If we were to add any sort of cleanup code at the end of the loop, it would be skipped altogether for `client_type == 1`. I have suggested this before and it wasn't an urgent issue back then either, but I find this style to be a gateway for very hard to find bugs.

File: bot/gwapi/facade.py (hunk 2 above)
Comment (Xyaren) [reply bot/gwapi/facade.py:87]: If paging parameters are not supplied, the response does not include the details
https://api.guildwars2.com/v2/characters?page=0&page_size=200&access_token=564F181A-F0FC-114A-A55D-3C1DCD45F3767AF3848F-AB29-4EBF-9594-F91E6A75E015&lang=en

https://api.guildwars2.com/v2/characters?access_token=564F181A-F0FC-114A-A55D-3C1DCD45F3767AF3848F-AB29-4EBF-9594-F91E6A75E015&lang=en

Taken from https://status.gw2efficiency.com/

File: bot/main.py (hunk 3 above)
Comment (Xyaren) [reply bot/main.py]: fixed

**Comment by Xyaren** [comment]
@ogrady As there have been some dramatic changes, I would like to request a re-review :)

File: bot/gwapi/facade.py (hunk 2 above)
Comment (ogrady) [reply bot/gwapi/facade.py:87]: Okay, so the number is basically arbitrary, got it.

File: bot/emblem_downloader.py [hunk 6]
```diff
@@ -10,11 +9,10 @@
...
 
 
-def download_guild_emblem(guild_id: str, guild_name: str, icon_size: int = 128) -> Tuple[Optional[int], Optional[bytes]]:
+def download_guild_emblem(guild_id: str, icon_size: int = 128) -> Optional[bytes]:
     """
     Download a guild emblem from emblem.werdes.net and checks if the response contains a real emblem or a fallback placeholder
     :param guild_id: ID of the guild as returned from the gw2 api
-    :param guild_name: Name of the guild, mainly used for the image_iud, which is actually a hash of the guilds name
     :param icon_size: Optional, size of the image. Anything over 128 is upscaled.
     :return: Icon ID, Icon image data or both null
```
Comment (ogrady) [inline bot/emblem_downloader.py:17]: Is this still correct?

File: bot/audit_service.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            LOG.info("User %s is no longer on our server. Removing access....", account_name)
+            self._user_service.removePermissions(client_unique_id)
+            self._user_service.remove_user_from_db(client_unique_id)
+
+    def trigger_user_audit(self):
+        LOG.info("Auditing users")
+        threading.Thread(name="FullAudit", target=self._audit_users, daemon=True).start()
+
+    def _audit_users(self):
+        self.c_audit_date = datetime.date.today()  # Update current date everytime run
```
Comment (ogrady) [inline bot/audit_service.py:85]: Linter does not like

File: bot/user_service.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    # Helps find the group ID for a group name
+    def _find_group_by_name(self, group_to_find):
+        with self._ts_connection_pool.item() as ts_facade:
+            self.groups_list = ts_facade.servergroup_list()
+        for group in self.groups_list:
+            if group.get('name') == group_to_find:
+                return group.get('sgid')
+        return -1
+
+    def clientNeedsVerify(self, unique_client_id):
```
Comment (ogrady) [inline bot/user_service.py:102]: Camelcase (also all following methods)

**Review by ogrady — COMMENTED** [review]
This review was brought to you by: my darn tablet with no English keyboard. :)

File: bot/ts/TS3Facade.py [hunk 9]
```diff
@@ -20,11 +21,21 @@ def __init__(self, ts3_connection: ThreadSafeTSConnection):
     def close(self):
         self._ts3_connection.close()
 
+    def __str__(self):
+        return f"TS3Facade[{self._ts3_connection}]"
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/ts/TS3Facade.py:25]: Log statement are using %s, for performance reasons, as the String is not built when not logged. 
This on the other hand will always be "formatted" when called.
f Style strings are much more readable in my opinion and should be favored if possible.

File: bot/ts/ThreadSafeTSConnection.py [hunk 10]
```diff
@@ -19,6 +19,11 @@ def default_exception_handler(ex):
     return ex
 
 
+def raise_exception_handler(ex):
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/ts/ThreadSafeTSConnection.py:22]: This would require manually raising the error through the whole callstack during the event loop. 
This is basically what Exceptions are made for.

File: bot/audit_service.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+                if ts_uuid is None:
+                    LOG.info("User %s is not found in TS DB and could be deleted.", audit_account_name)
+                    self._database_connection.cursor.execute("UPDATE users SET last_audit_date = ? WHERE ts_db_id= ?", ((datetime.date.today()), audit_ts_id,))
+                    # self.removeUserFromDB(audit_ts_id)
+                else:
+                    LOG.info("User %s is due for auditing! Queueing", audit_account_name)
+                    self.queue_user_audit(QUEUE_PRIORITY_AUDIT, audit_account_name, audit_api_key, audit_ts_id)
+
+        with self._database_connection.lock:
+            self._database_connection.cursor.execute('INSERT INTO bot_info (last_succesful_audit) VALUES (?)', (self.c_audit_date,))
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/audit_service.py:113]: I would leave it in for now, as it shows the last "queueing" for audit.

File: bot/event_looper.py [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+                self._broadcast_message()  # Send initial message into channel
+
+                # Forces script to loop forever while we wait for events to come in, unless connection timed out or exception occurs.
+                # Then it should loop a new bot into creation.
+                LOG.info("BOT now idle, waiting for requests.")
+
+                self._loop_for_events()
+
+    def _loop_for_events(self):
+        while self._ts_facade.is_connected():
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/event_looper.py:57]: I'm not 100% happy with the calss, as some class variables should probably be better handled as parameters in this call stack OR handled with instances of smaller scoped "Worker" Classes.

File: bot/event_looper.py [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+                                    if result is None:
+                                        LOG.debug("Failed. The group probably doesn't exist or the user is already hiding that group.")
+                                        self._ts_facade.send_text_message_to_client(rec_from_id, self._config.locale.get("bot_hide_guild_unknown"))
+                                    else:
+                                        guild_db_id = result[0]
+                                        self._database_connection.cursor.execute(
+                                            "INSERT INTO guild_ignores(guild_id, ts_db_id, ts_name) VALUES(?, ?, ?)",
+                                            (guild_db_id, rec_from_uid, rec_from_name))
+                                        self._database_connection.conn.commit()
+                                        LOG.debug("Success!")
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/event_looper.py:153]: In context with the previous message this is fine for me.

File: bot/event_looper.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+                        if len(args) == 1:
+                            LOG.info("User '%s' wants to unhide guild '%s'.", rec_from_name, args[0])
+                            with self._database_connection.lock:
+                                self._database_connection.cursor.execute(
+                                    "DELETE FROM guild_ignores WHERE guild_id = (SELECT guild_id FROM guilds WHERE ts_group = ? AND ts_db_id = ?)",
+                                    (args[0], rec_from_uid))
+                                changes = self._database_connection.cursor.execute("SELECT changes()").fetchone()[0]
+                                self._database_connection.conn.commit()
+                                if changes > 0:
+                                    LOG.debug("Success!")
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/event_looper.py:171]: same :)

File: bot/reset_roster_service.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+                pattern, clean = channels[i]
+                lead = leads[i]
+
+                shortened = StringShortener(TS3_MAX_SIZE_CHANNEL_NAME - len(clean)).shorten(lead)
+                newname = "%s%s" % (clean, ", ".join(shortened))
+
+                channel = facade.channel_find(pattern)
+                if channel is None:
+                    LOG.warning("No channel found with pattern '%s'. Skipping.", pattern)
+                    return
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/reset_roster_service.py:33]: should the broadcast be removed aswell ?
> Hello there! You can begin verification by typing 'verifyme' in this channel.

File: bot/ts/TS3Facade.py [hunk 16]
```diff
@@ -20,11 +21,21 @@ def __init__(self, ts3_connection: ThreadSafeTSConnection):
...
+
+    def version(self):
+        return self._ts3_connection.ts3exec_raise(lambda tc: tc.query("version").first())
 
     def wait_for_event(self, timeout: int):
-        return self._ts3_connection.ts3exec(lambda tc: tc.wait_for_event(timeout=timeout), ignore_exception_handler)[0]
+        try:
+            resp = self._ts3_connection.ts3exec_raise(lambda tc: tc.wait_for_event(timeout=timeout))
+        except TS3TimeoutError:
+            return None
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/ts/TS3Facade.py:37]: I'm much more of a fan of early returns, especially in a indentation based language like python.
It makes diffs much more readable for example if code gets added or removed.

Readup: https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement

File: bot/user_service.py (hunk 8 above)
Comment (Xyaren) [reply bot/user_service.py:102]: Those were just moved from the "Bot" Class.

Comment (Xyaren) [reply bot/user_service.py]: Fixed

File: bot/event_looper.py [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+                                    self._ts_facade.send_text_message_to_client(rec_from_id, self._config.locale.get("bot_unhide_guild_success"))
+                                else:
+                                    LOG.debug("Failed. Either the guild is unknown or the user had not hidden the guild anyway.")
+                                    self._ts_facade.send_text_message_to_client(rec_from_id, self._config.locale.get("bot_unhide_guild_unknown"))
+                        else:
+                            self._ts_facade.send_text_message_to_client(rec_from_id, self._config.locale.get("bot_unhide_guild_help"))
+                    elif cmd == 'verifyme':
```
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/event_looper.py]: should the broadcast be removed as well ?
> Hello there! You can begin verification by typing 'verifyme' in this channel.

File: bot/user_service.py (hunk 8 above)
Comment (Xyaren) [reply bot/user_service.py]: Fixed.

File: bot/audit_service.py (hunk 7 above)
Comment (Xyaren) [reply bot/audit_service.py]: Linter was right. Fixed

File: bot/emblem_downloader.py (hunk 6 above)
Comment (Xyaren) [reply bot/emblem_downloader.py]: Fixed

File: bot/event_looper.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ogrady) [reply bot/event_looper.py]: That would make sense then, yes.

File: bot/ts/TS3Facade.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ogrady) [reply bot/ts/TS3Facade.py]: This is arguably an early return in this particular case, as the whole function is just one statement (other than that, not really in line with SE-discussion), but same reasoning as before: putting cleanup code below `except` is then skipped altogether when catching an exception.

File: bot/ts/ThreadSafeTSConnection.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ogrady) [reply bot/ts/ThreadSafeTSConnection.py:22]: Absolutely! And I am not particularly a fan of raising exceptions through multiple layers of the call stack (read: `goto`), unless they signal absolutely application-breaking circumstances in which any cleanup can be bypassed (this probably ultimately lead to the invention of the `finally` block, which basically is spaghetti to me).

File: bot/event_looper.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ogrady) [reply bot/event_looper.py:57]: Understandably so. Since I have switched sides into the FP camp through the more recent years, I would advocate for passing all required information through as parameters to facilitate referential transparency, but I can see how that can be tedious for deep call stacks and we are in OOP after all. So maybe sparse use of singleton for those three classes is in order?

File: bot/ts/TS3Facade.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/ts/TS3Facade.py]: Cleanup code should be handled in a "finally" block or even better using a "with" statement.

Yes this is only one small function, but in terms of extendability the same rule applies here.

File: bot/ts/ThreadSafeTSConnection.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/ts/ThreadSafeTSConnection.py:22]: That is the case here. We are talking about Connection Problems, which can not be dealt with where they occur.

File: bot/event_looper.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/event_looper.py]: Done,

File: bot/event_looper.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Xyaren) [reply bot/event_looper.py:54]: Not sure how this would look like ?
Fow now I can add a NoneCheck in the while condition

[the remark this answers is not in the document]
Comment (ogrady) [reply bot/event_looper.py:54]: Yes, that's exactly what I had in mind.

**Comment by Xyaren** [comment]
Update:
Added SSH option (default is still telnet)

**Comment by Xyaren** [comment]
Pylint is currently restricted to Python 3.8.x (https://github.com/PyCQA/pylint/issues/3882#issuecomment-710780214)
I changed the build action to use 3.8.x instead of 3.x
